src/MongoDB.Bson/Serialization/Serializers/ObjectSerializer.cs

```diff
@@ -88,12 +102,28 @@ public override object Deserialize(BsonDeserializationContext context, BsonDeser
                     goto default;
 
                 case BsonType.Binary:
-                    var binaryData = bsonReader.ReadBinaryData();
-                    var subType = binaryData.SubType;
-                    if (subType == BsonBinarySubType.UuidStandard || subType == BsonBinarySubType.UuidLegacy)
+#pragma warning disable 618
+                    if (BsonDefaults.GuidRepresentationMode == GuidRepresentationMode.V2 && _guidRepresentation == GuidRepresentation.Unspecified)
                     {
-                        return binaryData.ToGuid();
+                        var binaryData = bsonReader.ReadBinaryData();
+                        var subType = binaryData.SubType;
+                        if (subType == BsonBinarySubType.UuidStandard || subType == BsonBinarySubType.UuidLegacy)
+                        {
+                            return binaryData.ToGuid();
+                        }
+                    }
+                    else
+                    {
+                        var binaryDataBookMark = bsonReader.GetBookmark();
```

Comment:
This is the new way, which delegates to GuidSerializer to do the actual deserialization. It is used when mode is V3 or a _guidRepresentation is specified.

Note the use of a bookmark so that we can read the same binary data twice, once here and then again in GuidSerializer.

Reply:
Book**m**ark?